### PR TITLE
New api routes for grade entry form controller

### PR DESCRIPTION
**Pull request by mishaschwartz**
- `grade_entry_forms_controller.create` : create a new grade entry form with optional columns
- `grade_entry_forms_controller.update` : update an existing grade entry form and update existing columns (or create new ones)
- these are required for the markus/quercus integration

### CODE REVIEW

File: app/controllers/api/grade_entry_forms_controller.rb [hunk 1]
```diff
@@ -31,6 +31,109 @@ def index
...
+    def update
+      # check if there is an existing assignment
+      form = GradeEntryForm.find_by_id(params[:id])
+      if form.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'Grade Entry Form not found' }, status: 404
+        return
+      end
+
+      ApplicationRecord.transaction do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply app/controllers/api/grade_entry_forms_controller.rb]: vote to ignore

File: app/controllers/api/grade_entry_forms_controller.rb [hunk 2]
```diff
@@ -31,6 +31,109 @@ def index
...
+
+    # create a new grade entry form
+    # params:
+    #   :short_identifier, :description, :date, :is_hidden
+    #   grade_entry_items:
+    #     :id, :name, :out_of, :bonus
+    #
+    # if the grade_entry_items id param is set, an existing item will be
+    # updated, otherwise a new grade_entry_item will be created
+    def update
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply app/controllers/api/grade_entry_forms_controller.rb]: vote to ignore

File: app/controllers/api/grade_entry_forms_controller.rb [hunk 3]
```diff
@@ -31,6 +31,109 @@ def index
...
+        new_form = GradeEntryForm.new(create_params)
+        unless new_form.save
+          render 'shared/http_status', locals: { code: '500', message:
+            new_form.errors.full_messages.first }, status: 500
+          raise ActiveRecord::Rollback
+        end
+
+        params[:grade_entry_items]&.each&.with_index do |column_params, i|
+          column_params = column_params.permit(:name, :out_of, :bonus).to_h.symbolize_keys
+          grade_item = GradeEntryItem.new(**column_params, grade_entry_form_id: new_form.id, position: i + 1)
```
Comment (david-yz-liu) [inline app/controllers/api/grade_entry_forms_controller.rb]: Prefer the association-based method here. Something like `new_form.grade_entry_items.build(...)`

File: app/controllers/api/grade_entry_forms_controller.rb [hunk 4]
```diff
@@ -31,6 +31,109 @@ def index
...
+      # check if there is an existing assignment
+      form = GradeEntryForm.find_by_id(params[:id])
+      if form.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'Grade Entry Form not found' }, status: 404
+        return
+      end
+
+      ApplicationRecord.transaction do
+        update_params = params.permit(*DEFAULT_FIELDS, :show_total)
```
Comment (david-yz-liu) [inline app/controllers/api/grade_entry_forms_controller.rb]: Allow the "is hidden" attribute too.

File: app/controllers/api/grade_entry_forms_controller.rb [hunk 5]
```diff
@@ -31,6 +31,109 @@ def index
...
+        unless form.update(update_params)
+          render 'shared/http_status', locals: { code: '500', message:
+            form.errors.full_messages.first }, status: 500
+          raise ActiveRecord::Rollback
+        end
+
+        params[:grade_entry_items]&.each do |column_params|
+          if column_params[:id].nil?
+            column_params = column_params.permit(:name, :out_of, :bonus).to_h.symbolize_keys
+            position = form.grade_entry_items.count + 1
```
Comment (david-yz-liu) [inline app/controllers/api/grade_entry_forms_controller.rb]: Initialize this once based on the count and then increment it manually in the loop (saves the extra DB queries).

File: app/controllers/api/grade_entry_forms_controller.rb (hunk 3 above)
Comment (david-yz-liu) [reply app/controllers/api/grade_entry_forms_controller.rb]: (and in `update` below)

File: app/controllers/api/grade_entry_forms_controller.rb (hunk 4 above)
Comment (mishaschwartz) [reply app/controllers/api/grade_entry_forms_controller.rb]: `is_hidden` is one of the `DEFAULT_FIELDS`

**Pull request merged**